### PR TITLE
Made changes in micropost controller to do nil check. #3

### PR DESCRIPTION
**Pull request by vimalathithanr**

### CODE REVIEW

**Comment by vimalathithanr** [comment]
@vnu Made changes to check nil. Can you please review it? 
I will wait till your review to merge and push the code to master.

File: app/controllers/microposts_controller.rb [hunk 1]
```diff
@@ -28,28 +28,31 @@ def show
...
-      favorite.micropost_id = params[:id]
-      favorite.user_id = current_user.id
-      favorite.save
-    rescue Exception => ex
-      flash[:error] = "#{ex}"
-      redirect_to request.referrer || root_url
-    end
+    
+      @favorite = Favorite.new
+      @favorite.micropost_id = params[:id]
```
Comment (vnu) [inline app/controllers/microposts_controller.rb]: This is not what I am expecting. 
Bug: If I pass in an invalid id, say 1000. Your code will blow with an exception. micropost with id 1000 not found!

**Comment by vnu** [comment]
@vimalathithanr This is how I would have done this
```ruby
before action :micropost, only: [:favorite, :unfavorite]

def favorite
  if micropost
    micropost.favorites.create(user_id: current_user.id)
    redirect_to request.referrer
  else
    flash.now[:error] = "Micropost with id: #{params[:id]} not found"
    render nothing: true, status: 404
  end
  
end

def unfavorite
  if micropost
    favorite = micropost.favorites.find_by(user_id: current_user.id)
    favorite ? favorite.destroy : flash.now[:error] = "Favorite not found"
    redirect_to request.referrer
  else
    flash.now[:error] = "Micropost with id: #{params[:id]} not found"
    render nothing: true, status: 404
  end
end

private

def micropost
  @micropost = Micropost.find_by(id: params[:id])
end
```

**Comment by vimalathithanr** [comment]
@vnu I understood your approach. This what I tried.

  def favorite
    
      @favorite = Favorite.new
      @favorite.micropost_id = params[:id]
      @favorite.user_id = current_user.id

      if @favorite.save
        redirect_to request.referrer || root_url
      else 
        flash.now[:error] = "Could not save"
        redirect_to request.referrer || root_url
      end

  end

  def unfavorite
      @favorite = Favorite.find_by(:user_id => current_user.id, :micropost_id => params[:id])

      if @favorite
        @favorite.destroy
        redirect_to request.referrer || root_url
      else 
        flash.now[:error] = "Could not Delete"
        redirect_to request.referrer || root_url
      end

  end

File: app/controllers/microposts_controller.rb (hunk 1 above)
Comment (vimalathithanr) [inline app/controllers/microposts_controller.rb]: @vnu I made changes. This is not what I did. Can you please go through the changes which I made.

Comment (vimalathithanr) [inline app/controllers/microposts_controller.rb]: def favorite
      @favorite = Favorite.new
      @favorite.micropost_id = params[:id]
      @favorite.user_id = current_user.id

      if @favorite.save
        redirect_to request.referrer || root_url
      else 
        flash.now[:error] = "Could not save"
        redirect_to request.referrer || root_url
      end

  end

  def unfavorite
      @favorite = Favorite.find_by(:user_id => current_user.id, :micropost_id => params[:id])

      if @favorite
        @favorite.destroy
        redirect_to request.referrer || root_url
      else 
        flash.now[:error] = "Could not Delete"
        redirect_to request.referrer || root_url
      end
  end

**Comment by vimalathithanr** [comment]
Made changes as suggested.

**Comment by vimalathithanr** [comment]
details
details1

File: app/controllers/microposts_controller.rb [hunk 2]
```diff
@@ -23,33 +24,33 @@ def destroy
...
-      favorite = Favorite.find_by(:user_id => current_user.id, :micropost_id => params[:id])
-      favorite.destroy
-    rescue Exception => ex
-      flash[:error] = "#{ex}"
-      redirect_to request.referrer || root_url
-    end
-          redirect_to request.referrer || root_url
+      if micropost
+        favorite = micropost.favorites.find_by(user_id: current_user.id)
+        if favorite
```
Comment (vnu) [inline app/controllers/microposts_controller.rb]: @vimalathithanr This favorite can be refactored
```ruby
if favorite
favorite.destroy
else
flash.now[:error] = "Favorite not found"
end
redirect_to request.referrer
```
Keep your code DRY (Don't repeat yourself)

**Pull request merged**